projects/storefrontlib/shared/components/hierarchy/hierarchy/hierarchy.component.service.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { OrderEntry, OrderEntryGroup } from '@spartacus/cart/base/root';
+import { HierarchyNode, Value } from '../hierarchy-node';
+import { TitleNode } from '../hierarchy-node-title';
+import { CollapsibleNode } from '../hierarchy-node-collapsible';
+import { Injectable } from '@angular/core';
+import { Observable, map } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+/**
+ * Service responsible for handling hierarchy-related operations.
+ */
+export class HierarchyComponentService {
+  /**
```

Comment:
@Zeyber  I would like to discuss with you `where` to put this file? 
First, the main functions of this service are:
1. Construct and prepare the data structure required by the `hierarchyComponent` component; 
2. Filter and construct `entries` of `standalone` type products from `entryGroups`.
The service needs to process and return `OrderEntry` and `OrderEntryGroup`, so it will inevitably reference `cart`.

Then, the reason why I put it in `storefrontlib/shared/component/hierarchy...` is mainly because every place that uses `hierarchy components` needs to call them, not just cart-related pages. I am also using it in `order, pickup-in-store, checkout/saved-cart, etc.` If I put it in one of the functions, such as the `cart directory`, then other functions will need to call the service in the cart when using it. I don't know but it doesn't sound very reasonable？

Finally, according to the idea of ​​making the `hierarchy` component more `dumb` and no any external reference. So it doesn't seem particularly appropriate to leave it here now.

So I would like to discuss this with you, do you have any suggestions on `where to put it `more appropriately, or any other ideas?

Reply:
I can see this component is not really related to the hierarchy component so much apart from the build tree function. IMO this is logic specific to the cart library in order to prepare your data for the hierarchy component. If you have some very generic helper method that can help build structure then it could live in a service like this. But I believe this is rather fine-grained to the cart component and should rather live somewhere there.